File: src/main/kotlin/lesson03/java/NumberUtil.java
package lesson03.java;

import java.util.List;

public class NumberUtil {
    private static NumberUtil sInstance;

    private NumberUtil() {

    }

    public static NumberUtil on() {
        if (sInstance == null) sInstance = new NumberUtil();
        return sInstance;
    }

    int addTwoNumbers(int firstNumber, int secondNumber) {
        return firstNumber + secondNumber;
    }

    void printNumberWithVoidReturnType(int number) {
        System.out.print(number);
        System.out.println();
    }

    void printNumber(int number) {
        System.out.print(number);
        System.out.println();
    }

    int sum(int... numbers) {
        int sum = 0;

        for (int currentNumber : numbers) {
            sum += currentNumber;
        }

        return sum;
    }

    int sum(List<Integer> numberList) {
        int sum = 0;

        for (int currentNumber : numberList) {
            sum += currentNumber;
        }

        return sum;
    }

    void max(int numberOne, int numberTwo) {
        NumberUtilCallback callback = System.out::println;
        callback.printNumber(numberOne >= numberTwo ? numberOne : numberTwo);
    }
}

File: src/main/kotlin/lesson03/kotlin/NumberUtil.kt
package lesson03.kotlin

object NumberUtil {
    fun addTwoNumbersWithSingleExpression(firstNumber: Int, secondNumber: Int): Int = firstNumber + secondNumber

    fun addTwoNumbers(firstNumber: Int, secondNumber: Int): Int {
        return firstNumber + secondNumber
    }

    fun printNumberWithUnitReturnType(number: Int = 0): Unit {
        print(number)
        println()
    }

    fun printNumber(number: Int = 0) {
        print(number)
        println()
    }

    fun sum(numberList: List<Int>, add: (Int, Int) -> Int): Int {
        var sum = 0

        for (currentNumber in numberList) {
            sum = add(sum, currentNumber)
        }

        return sum
    }

    fun sum(vararg numbers: Int): Int {
        var sum = 0

        for (currentNumber in numbers) {
            sum += currentNumber
        }

        return sum
    }

    fun sum(vararg numbers: Int, add: (Int, Int) -> Int): Int {
        var sum = 0

        for (currentNumber in numbers) {
            sum = add(sum, currentNumber)
        }

        return sum
    }

    fun max(numberOne: Int, numberTwo: Int) {
        fun printNumber(number: Int) {
            println(number)
        }

        printNumber(if (numberOne >= numberTwo) numberOne else numberTwo)
    }
}